### PR TITLE
Kddimitrov/restart on hmr fail

### PR DESCRIPTION
**Pull request by KristianDD**
## PR Checklist

- [ ] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [ ] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [ ] You have signed the [CLA](http://www.nativescript.org/cla).
- [ ] All existing tests are passing: https://github.com/NativeScript/nativescript-cli/blob/master/CONTRIBUTING.md#contribute-to-the-code-base
- [ ] Tests for the changes are included.

## What is the current behavior?

When HMR fails to apply patch the application will remain with old content.

## What is the new behavior?

When HMR fails to apply patch the application gets updated by regular LS and restarted.

Fixes/Implements/Closes [#3892](https://github.com/NativeScript/nativescript-cli/issues/3892).

### CODE REVIEW

File: lib/definitions/hmr-status-service.d.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+interface IHmrStatusService {
+	awaitHmrStatus(deviceId: string, operationHash: string): Promise<number>;
```
Comment (DimitarTachev) [inline lib/definitions/hmr-status-service.d.ts]: awaitHmrStatus -> getHmrStatus

File: lib/services/hmr-status-service.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+	public awaitHmrStatus(deviceId: string, operationHash: string): Promise<number> {
+		return new Promise((resolve, reject) => {
+			const key = `${deviceId}${operationHash}`;
+			let retryCount = 40;
+
+			const interval = setInterval(() => {
+				const status = this.getStatusByKey(key);
+				if (status || retryCount === 0) {
+					clearInterval(interval);
```
Comment (DimitarTachev) [inline lib/services/hmr-status-service.ts]: \+ clear on dispose

File: lib/services/livesync/playground/preview-app-livesync-service.ts [hunk 3]
```diff
@@ -43,19 +45,37 @@ export class PreviewAppLiveSyncService implements IPreviewAppLiveSyncService {
...
-					promise = this.applyChanges(this.$platformsData.getPlatformData(platform, projectData), projectData, filesToSyncMap[platform]);
+					const platformData = this.$platformsData.getPlatformData(platform, projectData);
+					const currentHmrData = _.cloneDeep(hmrData);
+					const filesToSync = _.cloneDeep(filesToSyncMap[platform]);
+					promise = this.applyChanges(platformData, projectData, filesToSync, data.appFilesUpdaterOptions.useHotModuleReload);
 					await promise;
+
+					if (data.appFilesUpdaterOptions.useHotModuleReload && currentHmrData.hash) {
+						const devices = _.filter(this.$previewSdkService.connectedDevices, { platform: platform.toLowerCase() });
+						_.forEach(devices, async (previewDevice: Device) => {
```
Comment (DimitarTachev) [inline lib/services/livesync/playground/preview-app-livesync-service.ts]: await properly here

File: lib/services/log-parser-service.ts [hunk 4]
```diff
@@ -23,22 +24,37 @@ export class LogParserService extends EventEmitter implements ILogParserService
 	@cache()
 	private startParsingLogCore(): void {
 		this.$deviceLogProvider.on(DEVICE_LOG_EVENT_NAME, (message: string, deviceIdentifier: string) => this.processDeviceLogResponse(message, deviceIdentifier));
+		this.$previewSdkService.on(DEVICE_LOG_EVENT_NAME, (message: string, deviceIdentifier: string) => this.processDeviceLogResponse(message, deviceIdentifier));
 	}
 
 	private processDeviceLogResponse(message: string, deviceIdentifier: string) {
-		const device = this.$devicesService.getDeviceByIdentifier(deviceIdentifier);
-		const devicePlatform = device.deviceInfo.platform.toLowerCase();
+		const devicePlatform = this.tryGetPlatform(deviceIdentifier);
```
Comment (DimitarTachev) [inline lib/services/log-parser-service.ts]: check the performance implications here (we will get the device or catch an exception for each log)

File: lib/services/livesync/playground/preview-app-livesync-service.ts [hunk 5]
```diff
@@ -43,19 +45,37 @@ export class PreviewAppLiveSyncService implements IPreviewAppLiveSyncService {
...
+							const status = await this.$hmrStatusService.awaitHmrStatus(previewDevice.id, currentHmrData.hash);
+							if (status === HmrConstants.HMR_ERROR_STATUS) {
+								await this.applyChanges(platformData, projectData, currentHmrData.fallbackFiles[platform], false);
+							}
```
Comment (Fatme) [inline lib/services/livesync/playground/preview-app-livesync-service.ts]: boolean variable passed as argument 🙀 🙀 🙀 
```
await this.applyChanges(platformData, projectData, currentHmrData.fallbackFiles[platform], { useHotModuleReload: false });
```

**Pull request merged**